azure-iot-device/azure/iot/device/common/pipeline/pipeline_stages_base.py

```diff
@@ -526,3 +526,184 @@ def _handle_pipeline_event(self, event):
+                logger.info("{}({}): returning timeout error".format(this.name, op.name))
+                this._complete_op(
+                    op,
+                    transport_exceptions.PipelineTimeoutError(
+                        "operation timed out before protocol client could respond"
+                    ),
+                )
+
+            logger.debug("{}({}): Creating timer".format(self.name, op.name))
+            op.timeout_timer = Timer(self.timeout_intervals[op.__class__], on_timeout)
+            op.timeout_timer.start()
+
+            # Send the op down, but intercept the return of the op so we can
+            # remove the timer when the op is done
+            logger.debug("{}({}): Sending down".format(self.name, op.name))
+            self._send_op_down_and_intercept_return(
+                op=op, intercepted_return=self._on_intercepted_return
+            )
+        else:
+            self._send_op_down(op)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _on_intercepted_return(self, op, error):
+        # When an op comes back, delete the timer and pass it right up.
+        del op.timeout_timer
+        logger.debug("{}({}): Op completed".format(self.name, op.name))
+        self._send_completed_op_up(op, error)
+
+
+class RetryStage(PipelineStage):
+    """
+    The purpose of the retry stage is to watch specific operations for specific
+    errors and retry the operations as appropriate.
+
+    This stage is currently very specific as it watches a hardcoded set of
+    operations for a hardcoded set of errors and retries after a hardcoded
+    period of time.
+
+    Like the TimeoutStage, the hardcoded nature of this stage will eventually
+    be provided by initialization parameters, and there will likely be two
+    retry stages which can handle IoTHub and MQTT retries at different levels
+    of the pipeline.
+
+    Unlike the TimeoutStage, this stage will never need to worry about cancelling
+    failed operations.  When an operation is retried at this stage, it is already
+    considered "failed", so no cancellation needs to be done.
+    """
+
+    def __init__(self):
+        super(RetryStage, self).__init__()
+        self.retry_intervals = {
+            pipeline_ops_mqtt.MQTTSubscribeOperation: 20,
+            pipeline_ops_mqtt.MQTTUnsubscribeOperation: 20,
+        }
+        self.waiting_to_retry = []
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _execute_op(self, op):
+        """
+        Send all ops down and intercept their return to "watch for retry"
+        """
+        self._send_op_down_and_intercept_return(
+            op=op, intercepted_return=self._on_intercepted_return
+        )
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def _should_watch_for_retry(self, op):
+        """
+        Return True if this op needs to be watched for retry.  This can be
+        called before the op runs.
+        """
+        return op.__class__ in self.retry_intervals
```

Comment:
I would appreciate this to be imported rather than defined on self

Reply:
Good point.  I'm thinking ahead to the time when retry_intervals will be an init param.  Leaving it here for now.

---
In reply to: [337155843](https://github.com/Azure/azure-iot-sdk-python/pull/365#discussion_r337155843) [](ancestors = 337155843)